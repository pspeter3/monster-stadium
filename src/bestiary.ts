import data from "./data.json";

export type SourceID =
    | "oota"
    | "pota"
    | "basic-rules"
    | "hotdq"
    | "mm"
    | "phb"
    | "skt"
    | "cos"
    | "tftyp"
    | "rot"
    | "vgtm";

export interface Source {
    readonly id: SourceID;
    readonly name: string;
}

export type Size =
    | "Tiny"
    | "Small"
    | "Medium"
    | "Large"
    | "Huge"
    | "Gargantuan";

export const Sizes = new Set<Size>([
    "Tiny",
    "Small",
    "Medium",
    "Large",
    "Huge",
    "Gargantuan",
]);

export type Alignment =
    | "Any"
    | "Any Chaotic"
    | "Any Evil"
    | "Chaotic Evil"
    | "Chaotic Good"
    | "Chaotic Neutral"
    | "Lawful Evil"
    | "Lawful Good"
    | "Lawful Neutral"
    | "Neutral"
    | "Neutral Evil"
    | "Neutral Good"
    | "Non-good"
    | "Non-lawful"
    | "Unaligned";

export const Alignments = new Set<Alignment>([
    "Any",
    "Any Chaotic",
    "Any Evil",
    "Chaotic Evil",
    "Chaotic Good",
    "Chaotic Neutral",
    "Lawful Evil",
    "Lawful Good",
    "Lawful Neutral",
    "Neutral",
    "Neutral Evil",
    "Neutral Good",
    "Non-good",
    "Non-lawful",
    "Unaligned",
]);

export type Environment =
    | "Aquatic"
    | "Arctic"
    | "Cave"
    | "Coast"
    | "Desert"
    | "Dungeon"
    | "Forest"
    | "Grassland"
    | "Mountain"
    | "Planar"
    | "Ruins"
    | "Swamp"
    | "Underground"
    | "Urban";

export const Environments = new Set<Environment>([
    "Aquatic",
    "Arctic",
    "Cave",
    "Coast",
    "Desert",
    "Dungeon",
    "Forest",
    "Grassland",
    "Mountain",
    "Planar",
    "Ruins",
    "Swamp",
    "Underground",
    "Urban",
]);

export type ChallengeRating =
    | 0
    | 0.125
    | 0.25
    | 0.5
    | 1
    | 2
    | 3
    | 4
    | 5
    | 6
    | 7
    | 8
    | 9
    | 10
    | 11
    | 12
    | 13
    | 14
    | 15
    | 16
    | 17
    | 18
    | 19
    | 20
    | 21
    | 22
    | 23
    | 24
    | 26
    | 30;

export const ChallengeRatings = new Set<ChallengeRating>([
    0,
    0.125,
    0.25,
    0.5,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    26,
    30,
]);

export interface Monster {
    readonly id: string;
    readonly name: string;
    readonly size: Size;
    readonly alignment: Alignment;
    readonly environments: ReadonlyArray<Environment>;
    readonly cr: ChallengeRating;
    readonly ac: number;
    readonly hp: number;
    readonly init: number;
    readonly legendary: boolean;
    readonly unique: boolean;
    readonly srd: false;
    readonly sources: ReadonlyArray<SourceID>;
}

export const Sources: ReadonlyArray<Source> = data.sources as Source[];
export const Monsters: ReadonlyArray<Monster> = data.monsters as Monster[];
