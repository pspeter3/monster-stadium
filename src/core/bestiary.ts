import data from "./data.json";

export type SourceID =
    | "basic-rules"
    | "cos"
    | "hotdq"
    | "mm"
    | "oota"
    | "phb"
    | "pota"
    | "rot"
    | "skt"
    | "srd"
    | "tftyp"
    | "vgtm";

export interface Source {
    readonly id: SourceID;
    readonly name: string;
}

export type Size =
    | "Tiny"
    | "Small"
    | "Medium"
    | "Large"
    | "Huge"
    | "Gargantuan";

export const Sizes = new Set<Size>([
    "Tiny",
    "Small",
    "Medium",
    "Large",
    "Huge",
    "Gargantuan",
]);

export type Type =
    | "Aberration"
    | "Beast"
    | "Celestial"
    | "Construct"
    | "Dragon"
    | "Elemental"
    | "Fey"
    | "Fiend"
    | "Giant"
    | "Humanoid"
    | "Monstrosity"
    | "Ooze"
    | "Plant"
    | "Undead";

export const Types = new Set<Type>([
    "Aberration",
    "Beast",
    "Celestial",
    "Construct",
    "Dragon",
    "Elemental",
    "Fey",
    "Fiend",
    "Giant",
    "Humanoid",
    "Monstrosity",
    "Ooze",
    "Plant",
    "Undead",
]);

export type Alignment =
    | "Any"
    | "Any Chaotic"
    | "Any Evil"
    | "Chaotic Evil"
    | "Chaotic Good"
    | "Chaotic Neutral"
    | "Lawful Evil"
    | "Lawful Good"
    | "Lawful Neutral"
    | "Neutral"
    | "Neutral Evil"
    | "Neutral Good"
    | "Non-good"
    | "Non-lawful"
    | "Unaligned";

export const Alignments = new Set<Alignment>([
    "Any",
    "Any Chaotic",
    "Any Evil",
    "Chaotic Evil",
    "Chaotic Good",
    "Chaotic Neutral",
    "Lawful Evil",
    "Lawful Good",
    "Lawful Neutral",
    "Neutral",
    "Neutral Evil",
    "Neutral Good",
    "Non-good",
    "Non-lawful",
    "Unaligned",
]);

export type Environment =
    | "Aquatic"
    | "Arctic"
    | "Cave"
    | "Coast"
    | "Desert"
    | "Dungeon"
    | "Forest"
    | "Grassland"
    | "Mountain"
    | "Planar"
    | "Ruins"
    | "Swamp"
    | "Underground"
    | "Urban";

export const Environments = new Set<Environment>([
    "Aquatic",
    "Arctic",
    "Cave",
    "Coast",
    "Desert",
    "Dungeon",
    "Forest",
    "Grassland",
    "Mountain",
    "Planar",
    "Ruins",
    "Swamp",
    "Underground",
    "Urban",
]);

export type ChallengeRating =
    | 0
    | 0.125
    | 0.25
    | 0.5
    | 1
    | 2
    | 3
    | 4
    | 5
    | 6
    | 7
    | 8
    | 9
    | 10
    | 11
    | 12
    | 13
    | 14
    | 15
    | 16
    | 17
    | 18
    | 19
    | 20
    | 21
    | 22
    | 23
    | 24
    | 26
    | 30;

export const ChallengeRatings = new Set<ChallengeRating>([
    0,
    0.125,
    0.25,
    0.5,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    26,
    30,
]);

export interface Monster {
    readonly id: string;
    readonly name: string;
    readonly size: Size;
    readonly type: Type;
    readonly alignment: Alignment;
    readonly environments: ReadonlyArray<Environment>;
    readonly cr: ChallengeRating;
    readonly hp: number;
    readonly ac: number;
    readonly init: number;
    readonly lair: boolean;
    readonly legendary: boolean;
    readonly unique: boolean;
    readonly sources: ReadonlyArray<SourceID>;
}

export type Bestiary = ReadonlyMap<string, Monster>;

export const Sources: ReadonlyArray<Source> = data.sources as Source[];
export const Monsters: Bestiary = new Map(
    data.monsters.map((monster) => [monster.id, monster as Monster])
);
