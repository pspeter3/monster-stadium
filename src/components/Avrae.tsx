import { h, FunctionComponent } from "preact";
import { Encounter } from "../hooks/useEncounter";
import { Bestiary } from "../core/bestiary";
import { useMemo } from "preact/hooks";

export const Avrae: FunctionComponent<{
    bestiary: Bestiary;
    encounter: Encounter;
}> = ({ bestiary, encounter }) => {
    const combat = useMemo(() => {
        const lines = [
            "!multiline",
            `!init begin dyn -name "${
                encounter.name !== undefined ? encounter.name : "Encounter"
            }"`,
        ];
        const prefixes = new Set<string>();
        Object.keys(encounter.monsters)
            .sort()
            .forEach((id) => {
                const monster = bestiary.get(id);
                if (monster) {
                    let prefix = "";
                    for (let i = 1; i <= id.length; i++) {
                        prefix = id.slice(0, i);
                        if (!prefixes.has(prefix)) {
                            prefixes.add(prefix);
                            break;
                        }
                    }
                    prefix = prefix.toUpperCase();
                    const count = encounter.monsters[id];
                    if (monster.sources.includes("srd")) {
                        lines.push(
                            `!init madd "${monster.name} -n ${count} -name "${prefix}# (${monster.name})" -rollhp -group ${id}`
                        );
                    } else {
                        for (let i = 1; i <= count; i++) {
                            lines.push(
                                `!init add ${monster.init} "${prefix}${i} (${monster.name})" -h -hp ${monster.hp} -ac ${monster.ac} -group ${id}`
                            );
                        }
                    }
                }
            });
        return lines.join("\n");
    }, [bestiary, encounter]);
    return (
        <pre class="bg-gray-800 leading-6 mx-4 overflow-x-scroll px-2 py-3 rounded text-gray-400">
            {combat}
        </pre>
    );
};
