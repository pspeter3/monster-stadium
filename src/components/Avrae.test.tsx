import { h } from "preact";
import { render, fireEvent } from "@testing-library/preact";
import { Avrae } from "./Avrae";
import { Monsters } from "../core/bestiary";

describe("Avrae", () => {
    it("should handle no name", () => {
        const ctx = render(
            <Avrae bestiary={Monsters} encounter={{ name: "", monsters: {} }} />
        );
        expect(ctx.baseElement).toMatchSnapshot();
    });

    it("should handle a name", () => {
        const ctx = render(
            <Avrae
                bestiary={Monsters}
                encounter={{ name: "Test", monsters: {} }}
            />
        );
        expect(ctx.baseElement).toMatchSnapshot();
    });

    it("should handle SRD monsters", () => {
        const ctx = render(
            <Avrae
                bestiary={Monsters}
                encounter={{ name: "", monsters: { "awakened-shrub": 2 } }}
            />
        );
        expect(ctx.baseElement).toMatchSnapshot();
    });

    it("should handle missing monsters", () => {
        const ctx = render(
            <Avrae
                bestiary={Monsters}
                encounter={{
                    name: "",
                    monsters: { "abominable-shrub": 2 },
                }}
            />
        );
        expect(ctx.baseElement).toMatchSnapshot();
    });

    it("should write to the clipboard", () => {
        const ctx = render(
            <Avrae
                bestiary={Monsters}
                encounter={{
                    name: "",
                    monsters: { "abominable-yeti": 2, "awakened-shrub": 2 },
                }}
            />
        );
        const mock = jest.fn();
        const clipboard = navigator.clipboard;
        (navigator as any).clipboard = {
            writeText: mock,
        } as any;
        fireEvent.click(ctx.getByText("Copy"));
        expect(mock).toHaveBeenCalled();
        (navigator as any).clipboard = clipboard;
    });
});
