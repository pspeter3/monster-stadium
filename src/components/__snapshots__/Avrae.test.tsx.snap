// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avrae should handle SRD monsters 1`] = `
<body>
  <div>
    <pre
      class="bg-gray-800 leading-6 mx-4 overflow-x-scroll px-2 py-3 rounded text-gray-400"
    >
      !multiline
!init begin dyn -name "Encounter"
!init madd "Awakened Shrub" -n 2 -name "A# (Awakened Shrub)" -rollhp -group awakened-shrub
    </pre>
    <div
      class="px-4 py-3"
    >
      <button
        class="dark:border-gray-700 dark:text-gray-500 dark:hover:bg-teal-900 dark:hover:text-teal-500 border-2 border-gray-200 focus:outline-none focus:shadow-outline focus:border-teal-600 font-bold hover:bg-teal-100 hover:border-teal-600 hover:text-teal-700 flex items-center justify-center h-12 rounded text-gray-700 w-full"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="13"
            rx="2"
            ry="2"
            width="13"
            x="9"
            y="9"
          />
          <path
            d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
          />
        </svg>
        <span
          class="leading-6 px-3 tracking-wide uppercase"
        >
          Copy
        </span>
      </button>
    </div>
  </div>
</body>
`;

exports[`Avrae should handle a name 1`] = `
<body>
  <div>
    <pre
      class="bg-gray-800 leading-6 mx-4 overflow-x-scroll px-2 py-3 rounded text-gray-400"
    >
      !multiline
!init begin dyn -name "Test"
    </pre>
    <div
      class="px-4 py-3"
    >
      <button
        class="dark:border-gray-700 dark:text-gray-500 dark:hover:bg-teal-900 dark:hover:text-teal-500 border-2 border-gray-200 focus:outline-none focus:shadow-outline focus:border-teal-600 font-bold hover:bg-teal-100 hover:border-teal-600 hover:text-teal-700 flex items-center justify-center h-12 rounded text-gray-700 w-full"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="13"
            rx="2"
            ry="2"
            width="13"
            x="9"
            y="9"
          />
          <path
            d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
          />
        </svg>
        <span
          class="leading-6 px-3 tracking-wide uppercase"
        >
          Copy
        </span>
      </button>
    </div>
  </div>
</body>
`;

exports[`Avrae should handle missing monsters 1`] = `
<body>
  <div>
    <pre
      class="bg-gray-800 leading-6 mx-4 overflow-x-scroll px-2 py-3 rounded text-gray-400"
    >
      !multiline
!init begin dyn -name "Encounter"
    </pre>
    <div
      class="px-4 py-3"
    >
      <button
        class="dark:border-gray-700 dark:text-gray-500 dark:hover:bg-teal-900 dark:hover:text-teal-500 border-2 border-gray-200 focus:outline-none focus:shadow-outline focus:border-teal-600 font-bold hover:bg-teal-100 hover:border-teal-600 hover:text-teal-700 flex items-center justify-center h-12 rounded text-gray-700 w-full"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="13"
            rx="2"
            ry="2"
            width="13"
            x="9"
            y="9"
          />
          <path
            d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
          />
        </svg>
        <span
          class="leading-6 px-3 tracking-wide uppercase"
        >
          Copy
        </span>
      </button>
    </div>
  </div>
</body>
`;

exports[`Avrae should handle name collisions 1`] = `
<body>
  <div>
    <pre
      class="bg-gray-800 leading-6 mx-4 overflow-x-scroll px-2 py-3 rounded text-gray-400"
    >
      !multiline
!init begin dyn -name "Encounter"
!init add 0 "A1 (Abominable Yeti)" -h -hp 137 -ac 15 -group abominable-yeti
!init add 0 "A2 (Abominable Yeti)" -h -hp 137 -ac 15 -group abominable-yeti
!init madd "Awakened Shrub" -n 2 -name "AW# (Awakened Shrub)" -rollhp -group awakened-shrub
    </pre>
    <div
      class="px-4 py-3"
    >
      <button
        class="dark:border-gray-700 dark:text-gray-500 dark:hover:bg-teal-900 dark:hover:text-teal-500 border-2 border-gray-200 focus:outline-none focus:shadow-outline focus:border-teal-600 font-bold hover:bg-teal-100 hover:border-teal-600 hover:text-teal-700 flex items-center justify-center h-12 rounded text-gray-700 w-full"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="13"
            rx="2"
            ry="2"
            width="13"
            x="9"
            y="9"
          />
          <path
            d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
          />
        </svg>
        <span
          class="leading-6 px-3 tracking-wide uppercase"
        >
          Copy
        </span>
      </button>
    </div>
  </div>
</body>
`;

exports[`Avrae should handle no name 1`] = `
<body>
  <div>
    <pre
      class="bg-gray-800 leading-6 mx-4 overflow-x-scroll px-2 py-3 rounded text-gray-400"
    >
      !multiline
!init begin dyn -name "Encounter"
    </pre>
    <div
      class="px-4 py-3"
    >
      <button
        class="dark:border-gray-700 dark:text-gray-500 dark:hover:bg-teal-900 dark:hover:text-teal-500 border-2 border-gray-200 focus:outline-none focus:shadow-outline focus:border-teal-600 font-bold hover:bg-teal-100 hover:border-teal-600 hover:text-teal-700 flex items-center justify-center h-12 rounded text-gray-700 w-full"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="13"
            rx="2"
            ry="2"
            width="13"
            x="9"
            y="9"
          />
          <path
            d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
          />
        </svg>
        <span
          class="leading-6 px-3 tracking-wide uppercase"
        >
          Copy
        </span>
      </button>
    </div>
  </div>
</body>
`;

exports[`Avrae should handle non SRD monters 1`] = `
<body>
  <div>
    <pre
      class="bg-gray-800 leading-6 mx-4 overflow-x-scroll px-2 py-3 rounded text-gray-400"
    >
      !multiline
!init begin dyn -name "Encounter"
!init add 0 "A1 (Abominable Yeti)" -h -hp 137 -ac 15 -group abominable-yeti
!init add 0 "A2 (Abominable Yeti)" -h -hp 137 -ac 15 -group abominable-yeti
    </pre>
    <div
      class="px-4 py-3"
    >
      <button
        class="dark:border-gray-700 dark:text-gray-500 dark:hover:bg-teal-900 dark:hover:text-teal-500 border-2 border-gray-200 focus:outline-none focus:shadow-outline focus:border-teal-600 font-bold hover:bg-teal-100 hover:border-teal-600 hover:text-teal-700 flex items-center justify-center h-12 rounded text-gray-700 w-full"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="13"
            rx="2"
            ry="2"
            width="13"
            x="9"
            y="9"
          />
          <path
            d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
          />
        </svg>
        <span
          class="leading-6 px-3 tracking-wide uppercase"
        >
          Copy
        </span>
      </button>
    </div>
  </div>
</body>
`;
